peps/pep-0750.rst

```diff
@@ -53,837 +33,1089 @@ deferred pending further experience with f-strings. Work on this PEP was
+processing code. For example, a template processor could cache the static parts
+of the template and only reprocess the dynamic parts when the template is
+evaluated with different values. Access to the static parts can be done with
+``template.args[::2]``.
 
-.. code-block:: python
 
-    html'<div id={id:int}>{content:HTML|str}</div>'
+Examples
+========
 
-In this case the ``format_spec`` for the second interpolation is the string
-``'HTML|str'``; it is up to the ``html`` tag to do something with the
-"format spec" here, if anything.
+All examples in this section of the PEP have fully tested reference implementations
+available in the public `pep750-examples <https://github.com/davepeck/pep750-examples>`_ 
+git repository.
 
-f-string-style ``=`` Evaluation
--------------------------------
 
-``mytag'{expr=}'`` is parsed to being the same as ``mytag'expr={expr}``', as
-implemented in the issue `Add = to f-strings for
-easier debugging <https://github.com/python/cpython/issues/80998>`_.
+Example: Implementing f-strings with t-strings
+----------------------------------------------
 
-Tag Function Arguments
-----------------------
+It is easy to "implement" f-strings using t-strings. That is, we can 
+write a function ``f(template: Template) -> str`` that processes a ``Template``
+in much the same way as an f-string literal, returning the same result:
 
-The tag function has the following signature:
 
 .. code-block:: python
 
-    def mytag(*args: Decoded | Interpolation) -> Any:
-        ...
+    name = "World"
+    value = 42
+    templated = t"Hello {name!r}, value: {value:.2f}"
+    formatted = f"Hello {name!r}, value: {value:.2f}"
+    assert f(templated) == formatted
 
-This corresponds to the following protocol:
+The ``f()`` function supports both conversion specifiers like ``!r`` and format
+specifiers like ``:.2f``. The full code is fairly simple:
 
 .. code-block:: python
 
-    class TagFunction(Protocol):
-        def __call__(self, *args: Decoded | Interpolation) -> Any:
-            ...
+    from types import Template, Interpolation
 
-Because of subclassing, the signature for ``mytag`` can of course be widened to
-the following, at the cost of losing some type specificity:
+    def convert(value: object, conv: Literal["a", "r", "s"] | None) -> object:
+        if conv == "a":
+            return ascii(value)
+        elif conv == "r":
+            return repr(value)
+        elif conv == "s":
+            return str(value)
+        return value
 
-.. code-block:: python
 
-    def mytag(*args: str | tuple) -> Any:
-        ...
+    def f(template: Template) -> str:
+        parts = []
+        for arg in template.args:
+            match arg:
+                case str() as s:
+                    parts.append(s)
+                case Interpolation(value, _, conv, format_spec):
+                    value = convert(value, conv)
+                    value = format(value, format_spec)
+                    parts.append(value)
```

Comment:
Perhaps simpler to do `if isinstance(arg, str) / else`? `match`/`case` seems a bit much, especially given that these are concrete types.

Reply:
Yeah. I'll defer to the community here as to what the most readable form is.

I think there's a general feeling that code that processes `Template` will often do interesting positional pattern matching, so it was probably worth including an example or two.